taichi/codegen/codegen_opengl.cpp

```diff
@@ -233,12 +235,43 @@ class KernelGen : public IRVisitor {
     }
 
     emit("int {} = {} + {} * {}; // {}", stmt->short_name(),
-         parent->short_name(),
-         struct_compiled_->class_children_map[parent_type],
+         parent->short_name(), struct_compiled_->class_children_map[parent_type],
          stmt->input_index->short_name(), stmt->snode->node_type_name);
   }
 
-  std::map<int, std::string> ptr_signats;
+  void visit(SNodeOpStmt *stmt) override {
+    auto sn = stmt->snode;
+    if (sn->type == SNodeType::bitmasked) {
+      auto ch_addr = fmt::format("_a_{}", stmt->short_name());
+      emit("int {} = {} + {} * {}; // {}", ch_addr, stmt->ptr->short_name(),
+          struct_compiled_->class_children_map[sn->node_type_name], stmt->val->short_name(),
+          sn->node_type_name);
+      if (stmt->op_type == SNodeOpType::is_active) {
+        emit("int {} = 1 & (_rt_.bitmask[{} >> 8] >> ({} & 31));",
+            stmt->short_name(), ch_addr, ch_addr);
+      } else if (stmt->op_type == SNodeOpType::activate) {
+        emit("atomicOr(_rt_.bitmask[{} >> 8], 1 << ({} & 31));", ch_addr, ch_addr);
+      } else if (stmt->op_type == SNodeOpType::deactivate) {
+        emit("atomicAnd(_rt_.bitmask[{} >> 8], ~(1 << ({} & 31)));", ch_addr, ch_addr);
+      }
+    } else if (sn->type == SNodeType::dense || sn->type == SNodeType::root) {
+      if (stmt->op_type == SNodeOpType::is_active) {
+        emit("int {} = 1;", stmt->short_name());
+      }
+    } else {
+        TI_NOT_IMPLEMENTED;
+    }
+  }
+
+  void visit(GlobalStoreStmt *stmt) override {
+    TI_ASSERT(stmt->width() == 1);
+    auto dt = stmt->data->element_type();
+    auto ch_addr = stmt->ptr->short_name();
+    emit("atomicOr(_rt_.bitmask[{} >> 8], 1 << ({} & 31));", ch_addr, ch_addr);
```

Comment:
Is this the correct usage to activate when store?

Reply:
Actually, `activation` should happen on `SNodeLookUp` statements. When doing `GlobalStore` you should assume the element is already activated.